dhnx/simulation.py

```diff
@@ -10,29 +10,751 @@
+
+                \lambda = 0.007 \cdot Re ^ {-0.13} \cdot D ^ {-0.14}
+
+            Parameters
+            ----------
+            re : pd.DataFrame
+                Reynoldes number for every timestep and pipe.
+
+            Returns
+            -------
+            lamb : pd.DataFrame
+                Darcy friction factor for every timestep and pipe.
+            """
+
+            factor_diameter = self.thermal_network.components.pipes[
+                ['from_node', 'to_node', 'diameter_mm']
+            ]
+
+            factor_diameter = (
+                1e-3 * factor_diameter.set_index(['from_node', 'to_node'])['diameter_mm']
+            ) ** -0.14
+
+            lamb = 0.07 * reynolds ** -0.13
+
+            lamb = lamb.multiply(factor_diameter, axis='columns')
+
+            return lamb
+
+        def _calculate_pipes_distributed_pressure_losses(lamb):
+            r"""
+            Calculates the pressure losses in the pipes.
+
+            .. math::
+
+                \delta p = \lambda \frac{8L}{\rho \pi^2 D^5}\dot{m}^2.
+
+            Parameters
+            ----------
+            lamb : pd.DataFrame
+                Darcy friction factor for every timestep and pipe.
+
+            Returns
+            -------
+            pipes_pressure_losses : pd.DataFrame
+                DataFrame with distributed pressure losses for every timestep and pipe.
+            """
+            pipes_mass_flow = self.results['pipes-mass_flow'].copy()
+
+            pipes_mass_flow_2 = pipes_mass_flow ** 2
+
+            constant = 8 * lamb / (self.rho * np.pi**2)
+
+            length = self.thermal_network.components.pipes[['from_node', 'to_node', 'length_m']]
+
+            diameter = \
+                self.thermal_network.components.pipes[['from_node', 'to_node', 'diameter_mm']]
+
+            length = length.set_index(['from_node', 'to_node'])['length_m']
+
+            diameter = diameter.set_index(['from_node', 'to_node'])['diameter_mm']
+
+            diameter_5 = (1e-3 * diameter) ** 5
+
+            pipes_pressure_losses = constant * pipes_mass_flow_2\
+                .multiply(length, axis='columns')\
+                .divide(diameter_5, axis='columns')
+
+            # We multiply by the factor of two to represent the pressure losses along inlet
+            # and return flow.
+
+            pipes_pressure_losses *= 2
+
+            return pipes_pressure_losses
+
+        def _calculate_pipes_localized_pressure_losses():
+            r"""
+            Calculates localized pressure losses at the nodes.
+
+            .. math::
+
+                \Delta p_{loc} = \frac{8\zeta\dot{m}^2}{\rho \pi^2 D^4}
+
+            Returns
+            -------
+            nodes_pressure_losses : pd.DataFrame
+                Pressure losses at the nodes.
+            """
+            def _assign_zeta_to_pipes(flow_type):
+
+                def func(row, level):
+
+                    res = zeta.loc[row.index.get_level_values(level).values]
```

Comment:
I receive a KeyError here: 'Passing list-likes to .loc or [] with any missing labels is no longer supported, see https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#deprecate-loc-reindex-listlike'

I tried fixing it using reindex() instead of loc. Unfortunately it did not work.

Reply:
"res = zeta.reindex(row.index.get_level_values(level).values)" works

Reply:
Thanks for pointing to that. I replaced it and it works. Solved?

Reply:
Yes, thank you! :)